bpi.syndicate.inc

```diff
@@ -543,3 +688,217 @@ function bpi_syndicate_action($bpi_id) {
+      $id = $name;
+      $type = $use_chosen ? 'select' : 'checkboxes';
+      $multiple = TRUE;
+      $options = array();
+      foreach ($terms as $term => $item) {
+        $options[$term] = $item['title'];
+      }
+      $default_value = _bpi_get_query_parameter(array('filter', $name), array());
+
+      switch ($name) {
+        case 'agency_internal':
+          $options = array(
+            '' => t('All'),
+          );
+          if (isset($terms[0])) {
+            $options['false'] = t('No');
+          }
+          if (isset($terms[1])) {
+            $options['true'] = t('Yes');
+          }
+          $default_value = _bpi_get_query_parameter(array('filter', $name));
+          $multiple = FALSE;
+          $type = 'select';
+          break;
+      }
+
+      $facets_form[$id] = array(
+        '#type' => $type,
+        '#multiple' => $multiple,
+        '#chosen' => $multiple && $use_chosen,
+        '#size' => $multiple ? min(20, count($options)) : NULL,
+        '#title' => bpi_label_mapper($name) ? bpi_label_mapper($name) : $name,
+        '#options' => $options,
+        '#default_value' => $default_value,
+        '#attributes' => array(
+          'data-placeholder' => $multiple ? t('Pick some values') : t('Pick a value'),
+          'data-no-matches' => t('Nothing matches'),
+        ),
+        '#prefix' => '<div class="bpi-facet bpi-facet-' . drupal_html_class($name) . '">',
+        '#suffix' => '</div>',
+      );
+    }
+
+    // Add logical operator.
+    $name = 'logicalOperator';
+    $id = $name;
+    $facets_form[$id] = array(
+      '#type' => 'select',
+      '#title' => bpi_label_mapper($name) ? bpi_label_mapper($name) : $name,
+      '#options' => array(
+        'and' => t('All must apply'),
+        'or' => t('Any must apply'),
+      ),
+      '#default_value' => _bpi_get_query_parameter(array('filter', $name), 'and'),
+      '#weight' => -10,
+    );
+
+    $filter_form['bpi_facets'] = $facets_form;
+
+    $filter_form['bpi_actions'] = array(
+      '#type' => 'container',
+      '#attributes' => array(
+        'class' => array(
+          drupal_html_class('container-inline'),
+        ),
+      ),
+    );
+
+    if (_bpi_get_current_filter()) {
+      $filter_form['bpi_actions']['reset_filter'] = array(
+        '#type' => 'item',
+        '#markup' => l(t('Reset filters'), current_path(), array(
+          'query' => _bpi_get_current_search_query(),
+          'attributes' => array(
+            'class' => array('button'),
+          ),
+        )),
+      );
+    }
+
+    $form['bpi_filter'] = $filter_form;
+  }
+}
+
+/**
+ * Get facets as an array.
+ *
+ * For performance reasons facets will only be loaded when no search or
+ * filters are applied.
+ *
+ * @return array
+ *   All facets.
+ */
+function bpi_get_facets() {
+  // Reload facets if they're not cached locally.
+  $reload = !isset($_SESSION['bpi_all_facets']) || !is_array($_SESSION['bpi_all_facets']);
```

Comment:
The data should be stored in the session. I've added a constant for the session key.